mart/attack/gain.py

```diff
@@ -84,3 +84,10 @@ def forward(self, rpn_objectness: torch.Tensor) -> torch.Tensor:
         else:
             # Encourage foreground.
             return probs
+
+
+from torch.nn import BCELoss as BCELoss_
+class BCELoss(BCELoss_):
+    def forward(self, input, target):
+        # We clone target because that tensor can be made in inference mode.
+        return super().forward(input, target.clone())
```

Comment:
This is a complete hack and needs to go away. The problem is anomalib creates tensors in inference_mode and we cannot use those tensors to compute a loss.

Reply:
Why can't we use those tensors to compute a loss? I wrote a minimal example that works.

```pycon
>>> import torch
>>> x = torch.tensor(3.0, requires_grad=True)
>>> 
>>> with torch.inference_mode():
...     y = torch.tensor(2.0)
... 
>>> (x + y).backward()
>>> assert x.grad == 1.0
>>> y._version
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: Inference tensors do not track version counter.
```

Reply:
What do you mean by "works"? The RuntimeError exception shows the problem.

Reply:
```
>>> import torch
>>> x = torch.tensor(3.0, requires_grad=True)
>>> with torch.inference_mode():
...   y = torch.tensor(2.0)
...
>>> torch.nn.functional.binary_cross_entropy(x, y)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/ccorneli/Projects/MART/venv/lib/python3.10/site-packages/torch/nn/functional.py", line 3122, in binary_cross_entropy
    return torch._C._nn.binary_cross_entropy(input, target, weight, reduction_enum)
RuntimeError: Inference tensors cannot be saved for backward. To work around you can make a clone to get a normal tensor and use it in autograd.
```

Reply:
You are right. I was not aware that PyTorch is a more restrictive on inference tensor in the CPP domain.

Reply:
They are reasonably specific that "Tensors created while the [inference] mode is enabled can[not] be used in grad-mode later": https://pytorch.org/docs/stable/notes/autograd.html#grad-modes.